dali/python/nvidia/dali/plugin/pytorch/experimental/proxy/__init__.py

```diff
@@ -0,0 +1,640 @@
+
+        # If it's a dictionary, apply the function to the values
+        elif isinstance(obj, dict):
+            return {key: self._produce_data_impl(value, cache) for key, value in obj.items()}
+
+        else:  # return directly anything else
+            return obj
+
+    def produce_data(self, obj):
+        """
+        A generic function to recursively visits all elements in a nested structure and replace
+        instances of DALIOutputBatchRef with the actual data provided by the DALI server
+
+        Args:
+            obj: The object to map (can be an instance of any class).
+
+        Returns:
+            A new object where any instance of DALIOutputBatchRef has been replaced with actual
+            data.
+        """
+        cache = dict()
+        ret = self._produce_data_impl(obj, cache)
+        del cache
+        return ret
+
+    def _get_input_batches(self, max_num_batches, timeout=None):
+        _torch.cuda.nvtx.range_push("dali_input_q.get")
+        count = 0
+        batches = []
+        if timeout is not None:
+            try:
+                batches.append(self._dali_input_q.get(timeout=timeout))
+                count = count + 1
+            except Empty:
+                return None
+            except _mp.TimeoutError:
+                return None
+
+        while count < max_num_batches:
+            try:
+                batches.append(self._dali_input_q.get_nowait())
+                count = count + 1
+            except Empty:
+                break
+        _torch.cuda.nvtx.range_pop()
+        return batches
+
+    def _thread_fn(self):
+        """
+        Asynchronous DALI thread that gets iteration data from the queue and schedules it
+        for execution
+        """
+        self._pipe.build()  # just in case
+
+        fed_batches = queue.Queue()
+        while not self._thread_stop_event.is_set():
+            _torch.cuda.nvtx.range_push("get_input_batches")
+            timeout = 5 if fed_batches.empty() else None
+            # We try to feed as many batches as the prefetch queue (if available)
+            batches = self._get_input_batches(
+                self._pipe.prefetch_queue_depth - fed_batches.qsize(), timeout=timeout
+            )
+            _torch.cuda.nvtx.range_pop()
+            if batches is not None and len(batches) > 0:
+                _torch.cuda.nvtx.range_push("feed_pipeline")
+                for batch_id, inputs in batches:
+                    for input_name, input_data in inputs.items():
+                        self._pipe.feed_input(input_name, input_data)
+                    self._pipe._run_once()
+                    fed_batches.put(batch_id)
+                _torch.cuda.nvtx.range_pop()
+
+            # If no batches to consume, continue
+            if fed_batches.qsize() == 0:
+                continue
+
+            _torch.cuda.nvtx.range_push("outputs")
+            batch_id = fed_batches.get_nowait()  # we are sure there's at least one
```

Comment:
just making sure, this is a local variable that is not shared elsewhere, right? Do we need to use a synchronized queue here?

Reply:
True. Making it a simple list